### PR TITLE
Add episode 36

### PR DESCRIPTION
**Pull request by bitsondatadev**

### CODE REVIEW

File: _episodes/36.md [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+     time: 0
+   - title: "Pull requests of the episode"
+     desc: ""
+     time: 0
+   - title: "Demo of the episode"
+     desc: ""
+     time: 0
+---
+
+## Releases 379 to 380
```
Comment (mosabua) [inline _episodes/36.md:29]:
```
## Releases 379 to 381

Official highlights from Martin Traverso:

[Trino 379](https://trino.io/docs/current/release/release-379.html)

* New MariaDB connector
* Performance improvements for `JOIN`, `UNION`, and `GROUP BY`
* Support for Google Cloud Storage in the Delta Lake connector
* Support for Pinot 0.10

[Trino 380](https://trino.io/docs/current/release/release-380.html)

* Update Cassandra connector to support v5 and v6 protocols.
* Rename properties controlling Hive view parsing.
* Allow changing file and table format with the Iceberg connector.
* Add support for bulk data insertion in SQL Server connector.

[Trino 381](https://trino.io/docs/current/release/release-381.html)

* Support for `UPDATE` in Iceberg connector.
* Experimental support for table functions.
* Support for exchange spooling on Azure Blob Storage.
* Support reading snapshot tables and materialized views in BigQuery connector.

Additional highlights worth a mention according to Manfred:

* Next is exchange spooling on [Google Cloud Storage](https://github.com/trinodb/trino/pull/12360).
* Framework for table functions is in place, implementations in connectors are coming.
* `ldap.ssl-trust-certificate` as legacy config removes upgrade failures.
* Introduce the `least-waste` low memory task killer policy.
* Disable auto-suggestion in CLI

More detailed information is available in the release notes for
[Trino 379](https://trino.io/docs/current/release/release-379.html),
[Trino 380](https://trino.io/docs/current/release/release-380.html),
and
[Trino 381](https://trino.io/docs/current/release/release-381.html).
```

Comment (mosabua) [inline _episodes/36.md:29]: 381

File: _episodes/36.md [hunk 2]
```diff
@@ -0,0 +1,191 @@
+---
+layout: episode
+title:  "36: Trino vectorizes and moves to Java 17"
+date: 2022-05-12
+tags: java17 vectorization garbage_collection performance
+youtube_id: ""
+wistia_id: ""
+sections:
+   - title: "Releases 379 and 380"
```
Comment (mosabua) [inline _episodes/36.md:9]: 381

File: _episodes/36.md [hunk 3]
```diff
@@ -0,0 +1,525 @@
...
+    if (obj instanceof String s) {
+        // Let pattern matching do the work!
+        ...
+    }
+    ```
+
+* [Helpful NullPointerExceptions](https://openjdk.java.net/jeps/358) are
+particularly exciting as the ever confusing nulls for no reason don't come up 
+and require you to chase down where it happened in the code. Instead there is
+new information added to the message that ideally will give you a more unique
```
Comment (mosabua) [inline _episodes/36.md:368]: ideally gives you

File: _episodes/36.md [hunk 4]
```diff
@@ -0,0 +1,525 @@
...
+message to throw at Stack Overflow.
+
+
+### Rearchitecting Trino
+
+With all of these exciting changes, what does this mean for Trino? Let's first
+dive into the thing that many of our users dread...upgrading.
+
+#### Upgrade to Java 17 (When  it's time)
```
Comment (mosabua) [inline _episodes/36.md:377]: remove duplicate space character

File: _episodes/36.md [hunk 5]
```diff
@@ -0,0 +1,525 @@
...
+
+1. Updating Java in all the clients and code that calls Trino is tedious. 
+
+    > You luckily only need to update the server that Trino is running on. The
+    client or CLI can still run any version of Java.
+                                                                              >
+2. There are conflicting Java versions on the node I am running.
+
+    > If you are running another application depending on Java you shouldn't be.
+    Ideally Trino will run on its own server. If there's a smaller application
```
Comment (mosabua) [inline _episodes/36.md:394]: Trino runs

Comment (mosabua) [reply _episodes/36.md:394]: and maybe "servers"

File: _episodes/36.md [hunk 6]
```diff
@@ -0,0 +1,525 @@
...
+    >You can have side-by-side installs of multiple versions of the JDK and use 
+    the appropriate one. You just need to launch Trino with the correct Java  
+    command. If your company is against using a newer JDK, you can point out the
+    arguments above to update the policy to at least include JDK17.
+ 
+
+
+#### Revamping Trino
+
+We're also in the process of revamping the core execution engine, which will
```
Comment (mosabua) [inline _episodes/36.md:410]: revamping is a bit much .. we are iterating and improving

File: _episodes/36.md [hunk 7]
```diff
@@ -0,0 +1,525 @@
...
+automatically and generate better machine code.
+* Implement evaluation of certain operations more efficiently by taking 
+advantage of SIMD or GPU-based processing.
+* Columnar evaluation.
+
+#### Project Hummingbird
+
+Just as we did with the efforts around 
+[Project Tardigrade](https://trino.io/blog/2022/05/05/tardigrade-launch.html) we
+want to centralize these efforts under a project name that will include a set
```
Comment (mosabua) [inline _episodes/36.md:424]: that includes 

also .. contributors, testers and others ?

File: _episodes/36.md [hunk 8]
```diff
@@ -0,0 +1,525 @@
...
+of motivated contributors and give it a cool name.
+
+After some discussion, we would like to announce Project Hummingbird will be the
```
Comment (mosabua) [inline _episodes/36.md:427]: "Project Hummingbird" is the new banner name for the

File: _episodes/36.md [hunk 9]
```diff
@@ -0,0 +1,525 @@
...
+banner for the efforts around improving performance and concentrated updates to
+core
```
Comment (mosabua) [inline _episodes/36.md:429]: to the core of Trino.

File: _episodes/36.md [hunk 10]
```diff
@@ -0,0 +1,525 @@
...
+Just as we did with the efforts around 
+[Project Tardigrade](https://trino.io/blog/2022/05/05/tardigrade-launch.html) we
+want to centralize these efforts under a project name that will include a set
+of motivated contributors and give it a cool name.
+
+After some discussion, we would like to announce Project Hummingbird will be the
+banner for the efforts around improving performance and concentrated updates to
+core 
+
+We chose Hummingbirds because they are adaptive, light, and fast. Hummingbirds are the only bird
```
Comment (mosabua) [inline _episodes/36.md:431]: hummingbirds as mascots

File: _episodes/36.md [hunk 11]
```diff
@@ -0,0 +1,525 @@
...
+with the incredible capability to fly in any direction and are super fast. It
+made sense as Trino evolves into a query engine that is capable of adapting to
+its environment during query runtime, it is akin to this agile and beautiful
```
Comment (mosabua) [inline _episodes/36.md:434]: these agile and beautiful creatures.

**Review by mosabua — CHANGES_REQUESTED** [review]
First round of review in..

File: _episodes/36.md [hunk 12]
```diff
@@ -0,0 +1,534 @@
...
+Usain Bolt in a few races. The two are so close in speed, it's seconds to 
+milliseconds difference. The main difference in this analogy is that speed is
+the only thing that matters in an olymic race. Howver, programming languages and
+frameworks have a lot more tradeoffs.
+
+<p align="center">
+<img align="center" width="75%" height="100%" src="/assets/episode/36/usain-bolt-yohan-blake.webp"/>
+</p>
+
+The point is, Java is fast and more importantly, it abstracts a lot of burden
```
Comment (mosabua) [inline _episodes/36.md:124]: probably better to use

it removes a lot of burden

Comment (mosabua) [inline _episodes/36.md:124]: probably better to use

it removes a lot of burden

File: _episodes/36.md [hunk 13]
```diff
@@ -0,0 +1,534 @@
...
+maintaining and scaling out the code. This is conducive to a healthy open-source
+project and lowers the barrier for collaboration. Rather than go against this
```
Comment (mosabua) [inline _episodes/36.md:126]: project, and

Comment (mosabua) [inline _episodes/36.md:126]: project, and

Comment (mosabua) [reply _episodes/36.md:126]: or project since it lowers

Comment (mosabua) [reply _episodes/36.md:126]: or project since it lowers

File: _episodes/36.md [hunk 14]
```diff
@@ -0,0 +1,534 @@
...
+<p align="center">
+<img align="center" width="75%" height="100%" src="/assets/episode/36/usain-bolt-yohan-blake.webp"/>
+</p>
+
+The point is, Java is fast and more importantly, it abstracts a lot of burden
+maintaining and scaling out the code. This is conducive to a healthy open-source
+project and lowers the barrier for collaboration. Rather than go against this 
+and take on the feat of having to rewrite an entire system in C++, why not lean
+into the incredible innovation recent Java features have to offer to produce
+speed.
```
Comment (mosabua) [inline _episodes/36.md:129]: to improve performance even more.

Comment (mosabua) [inline _episodes/36.md:129]: to improve performance even more.

File: _episodes/36.md [hunk 15]
```diff
@@ -0,0 +1,534 @@
...
+constantly being added to every release that ultimately makes for a 
+work-smarter-not-harder query engine.
+
+## Concept of the episode: Java 17 and rearchitecting Trino
+
+As Trino prepares to [update to Java 17](https://github.com/trinodb/trino/issues/9876),
+we wanted to give a glimpse at what has happened between the current required
+JDK version, JDK 11, and future version JDK 17. Both of these versions are
+long-term support versions, and in the four years from 11 to 17 
+[a lot of exciting improvements were added.](https://openjdk.java.net/projects/jdk/17/jeps-since-jdk-11).
```
Comment (mosabua) [inline _episodes/36.md:151]: were added](


remove . inside ]

File: _episodes/36.md [hunk 16]
```diff
@@ -0,0 +1,534 @@
...
+
+### Java 17 updates
+
+Here are some [updates coming up in Java 17](https://openjdk.java.net/projects/jdk/17/jeps-since-jdk-11).
+
+#### Performance
+
+There were several JEPS that improve performance as well as many small changes
```
Comment (mosabua) [inline _episodes/36.md:159]: Java Enhancement Proposals (JEP)

Comment (mosabua) [inline _episodes/36.md:159]: Java Enhancement Proposals (JEP)

Comment (mosabua) [reply _episodes/36.md:159]: also confirm that this is what JEP stands for

File: _episodes/36.md [hunk 17]
```diff
@@ -0,0 +1,534 @@
...
+* [JEP 352](https://openjdk.java.net/jeps/352)
+* [JEP 356](https://openjdk.java.net/jeps/356)
+* [JEP 387](https://openjdk.java.net/jeps/387)
+* [JEP 412](https://openjdk.java.net/jeps/412)
+
+Performance is a multifaceted topic that includes factors like throughput, 
+latency, memory footprint, startup, ramp up, pause times, and shut down time.
+
+You can used standardized benchmarks like 
+[SPECjbb® 2015](https://www.spec.org/jbb2015/) to test Java application in most
```
Comment (mosabua) [inline _episodes/36.md:172]: to test a Java application

Comment (mosabua) [inline _episodes/36.md:172]: to test a Java application

File: _episodes/36.md [hunk 18]
```diff
@@ -0,0 +1,534 @@
...
+Martin did a similar test late last year, and reported anywhere from 10-15% 
+improvement in speed in Java 17 using the G1 garbage collector. This is an 
+exciting development and we hope to publish more about this as we get closer to 
+updating.
+
+#### Garbage collectors
+
+Although garbage collectors are performance enhancements in their own right, 
+there is a lot of exciting changes around garbage collectors in Java 17 since 
+Java 11 that makes gives them their own spot.
```
Comment (mosabua) [inline _episodes/36.md:189]: that makes gives ???

Comment (mosabua) [inline _episodes/36.md:189]: that makes gives ???

File: _episodes/36.md [hunk 19]
```diff
@@ -0,0 +1,534 @@
...
+
+First not one, but two concurrent garbage collectors have made their way out of
+incubation, and ready.
```
Comment (mosabua) [inline _episodes/36.md:192]: , and are ready for use:

Comment (mosabua) [inline _episodes/36.md:192]: , and are ready for use:

File: _episodes/36.md [hunk 20]
```diff
@@ -0,0 +1,534 @@
...
+</p>
+
+<p align="center">
+<img align="center" width="75%" height="100%" src="/assets/episode/36/footprint.png"/><br/>
+Source: <a href="https://kstefanj.github.io/2021/11/24/gc-progress-8-17.html">Stefan Johansson's Blog</a>
+</p>
+
+Take particular note of the massive improvement of G1. This is especially 
+exciting because G1 is recommended for Trino usage. It's still too early to 
+claim that ZGC or Shenendoah will have overall better performance depending on
```
Comment (mosabua) [inline _episodes/36.md:235]: to determine if ZGC ..

File: _episodes/36.md [hunk 21]
```diff
@@ -0,0 +1,534 @@
...
+without manual intervention.
+
+However, this isn't always so straightforward to detect. As you can see from the
+comments in the example, special criteria need to be met. For this, you can use
+the Vector API to directly interface with SIMD and GPU instructions. We will 
+show more on this in the demo.
+
+#### Language features
+
+Beyond the performance improvements Java 17 includes some exciting new Java
```
Comment (mosabua) [inline _episodes/36.md:291]: improvements, Java 17

Comment (mosabua) [inline _episodes/36.md:291]: improvements, Java 17

File: _episodes/36.md [hunk 22]
```diff
@@ -0,0 +1,534 @@
...
+However, this isn't always so straightforward to detect. As you can see from the
+comments in the example, special criteria need to be met. For this, you can use
+the Vector API to directly interface with SIMD and GPU instructions. We will 
+show more on this in the demo.
+
+#### Language features
+
+Beyond the performance improvements Java 17 includes some exciting new Java 
+language updates and improvements. While some may not consider this as exciting
+as performance boosts, Language enhancements make it easier to write more
```
Comment (mosabua) [inline _episodes/36.md:293]: language

Comment (mosabua) [inline _episodes/36.md:293]: language

File: _episodes/36.md [hunk 23]
```diff
@@ -0,0 +1,534 @@
...
+comments in the example, special criteria need to be met. For this, you can use
+the Vector API to directly interface with SIMD and GPU instructions. We will 
+show more on this in the demo.
+
+#### Language features
+
+Beyond the performance improvements Java 17 includes some exciting new Java 
+language updates and improvements. While some may not consider this as exciting
+as performance boosts, Language enhancements make it easier to write more 
+quality and maintainable code. This is especially important for an open source
```
Comment (mosabua) [inline _episodes/36.md:294]: higher quality and

Comment (mosabua) [inline _episodes/36.md:294]: higher quality and

File: _episodes/36.md [hunk 24]
```diff
@@ -0,0 +1,534 @@
...
+quality and maintainable code. This is especially important for an open source 
+project that is maintained by many individuals.
+
+* A very useful change for Trino is the new support for 
+[multiline text blocks](https://openjdk.java.net/jeps/378). This allows you to 
+go from having to write a SQL query represented in a one-dimensional string 
+literal like this:
+
+    ```
+    String query = "SELECT \"EMP_ID\", \"LAST_NAME\" FROM \"EMPLOYEE_TB\"\n" +
```
Comment (mosabua) [inline _episodes/36.md:303]: I think we should change field names to lowercase here and in the next snippet

Comment (mosabua) [inline _episodes/36.md:303]: I think we should change field names to lowercase here and in the next snippet

File: _episodes/36.md [hunk 25]
```diff
@@ -0,0 +1,534 @@
...
+    ```
+    switch (day) {
+        case MONDAY, FRIDAY, SUNDAY -> System.out.println(6);
+        case TUESDAY                -> System.out.println(7);
+        case THURSDAY, SATURDAY     -> System.out.println(8);
+        case WEDNESDAY              -> System.out.println(9);
+    }
+    ```
+* Always having to do a cast after checking for a type has always been an
+annoyance to many Java developers, especially for many types.
```
Comment (mosabua) [inline _episodes/36.md:353]: can we drop ", especially for many types."

Comment (mosabua) [inline _episodes/36.md:353]: can we drop ", especially for many types."

File: _episodes/36.md [hunk 26]
```diff
@@ -0,0 +1,534 @@
...
+    
+    ```
+    if (obj instanceof String s) {
+        // Let pattern matching do the work!
+        ...
+    }
+    ```
+
+* [Helpful NullPointerExceptions](https://openjdk.java.net/jeps/358) are
+particularly exciting as the ever confusing nulls for no reason don't come up
```
Comment (mosabua) [inline _episodes/36.md:375]: come up, and

Comment (mosabua) [inline _episodes/36.md:375]: come up, and

File: _episodes/36.md [hunk 27]
```diff
@@ -0,0 +1,534 @@
...
+        // Let pattern matching do the work!
+        ...
+    }
+    ```
+
+* [Helpful NullPointerExceptions](https://openjdk.java.net/jeps/358) are
+particularly exciting as the ever confusing nulls for no reason don't come up 
+and require you to chase down where it happened in the code. Instead there is
+new information added to the message that ideally gives you a more unique
+message to throw at Stack Overflow.
```
Comment (mosabua) [inline _episodes/36.md:378]: remove "to throw at Stack Overflow"

Comment (mosabua) [inline _episodes/36.md:378]: remove "to throw at Stack Overflow"

File: _episodes/36.md [hunk 28]
```diff
@@ -0,0 +1,534 @@
...
+
+
+### Rearchitecting Trino
+
+With all of these exciting changes, what does this mean for Trino? Let's first
```
Comment (mosabua) [inline _episodes/36.md:383]: With all these

Comment (mosabua) [inline _episodes/36.md:383]: With all these

File: _episodes/36.md [hunk 29]
```diff
@@ -0,0 +1,534 @@
...
+dive into the thing that many of our users dread...upgrading.
+
+#### Upgrade to Java 17 (When it's time)
+ 
+As mentioned before, Java 17 is the next LTS verison since Java 11. Java 17
```
Comment (mosabua) [inline _episodes/36.md:388]: is the current LTS version, following Java 11.

Comment (mosabua) [inline _episodes/36.md:388]: is the current LTS version, following Java 11.

File: _episodes/36.md [hunk 30]
```diff
@@ -0,0 +1,534 @@
...
+we update, everyone should be running version 17 to get the best experience out
+of Trino. Moving to Java 17 allows us to take advantage of many improvements to
+the JDK and the Java language that were introduced since Java 11. There are 
+some reasons people say they can't update.
+
+1. Updating Java in all the clients and code that calls Trino is tedious. 
+
+    > You luckily only need to update the server that Trino is running on. The
+    client or CLI can still run any version of Java.
+                                                                              >
```
Comment (mosabua) [inline _episodes/36.md:399]: remove >

Comment (mosabua) [inline _episodes/36.md:399]: remove >

File: _episodes/36.md [hunk 31]
```diff
@@ -0,0 +1,534 @@
...
+2. There are conflicting Java versions on the node Trino servers run.
+
+    > If you are running another application depending on Java you shouldn't be.
+    Ideally Trino will run on its own server. If there's a smaller application
+    to, for example, monitor Trino, then you should be able to install a
+    seperate version of Trino.
+                                                                    >
```
Comment (mosabua) [inline _episodes/36.md:406]: remove >

File: _episodes/36.md [hunk 32]
```diff
@@ -0,0 +1,534 @@
...
+3. There is a company policy requiring specific JDKs be installed on all 
+    servers. 
+    
+    >You can have side-by-side installs of multiple versions of the JDK and use 
+    the appropriate one. You just need to launch Trino with the correct Java  
+    command. If your company is against using a newer JDK, you can point out the
+    arguments above to update the policy to at least include JDK17.
+
```
Comment (mosabua) [inline _episodes/36.md:414]: Remove redundant empty lines

Comment (mosabua) [inline _episodes/36.md:414]: Remove redundant empty lines

File: _episodes/36.md [hunk 33]
```diff
@@ -0,0 +1,534 @@
...
+    the appropriate one. You just need to launch Trino with the correct Java  
+    command. If your company is against using a newer JDK, you can point out the
+    arguments above to update the policy to at least include JDK17.
+ 
+
+
+#### Iterating and improving Trino
+
+We're also in the process of revamping the core execution engine, which 
+enables us to:
```
Comment (mosabua) [inline _episodes/36.md:420]: to implement the following improvements: 

and empty line before list

**Review by mosabua — CHANGES_REQUESTED** [review]

File: _episodes/36.md (hunk 16 above)
Comment (bitsondatadev) [reply _episodes/36.md:159]: actually it's JDK enhancement proposals

File: _episodes/36.md [hunk 34]
```diff
@@ -0,0 +1,533 @@
...
+long-term support versions, and in the four years from 11 to 17 
+[a lot of exciting improvements were added](https://openjdk.java.net/projects/jdk/17/jeps-since-jdk-11).
+
+### Java 17 updates
+
+Here are some [updates coming up in Java 17](https://openjdk.java.net/projects/jdk/17/jeps-since-jdk-11).
+
+#### Performance
+
+There were several JDK Enhancement Proposals (JEPS) that improve performance as
```
Comment (mosabua) [inline _episodes/36.md:161]: (JEP)

File: _episodes/36.md (hunk 5 above)
Comment (mosabua) [reply _episodes/36.md:404]: ping

File: _episodes/36.md [hunk 35]
```diff
@@ -0,0 +1,533 @@
...
+[Trino 381](https://trino.io/docs/current/release/release-381.html).
+
+### Cinco de Trino recap blog post
+
+Check out this blog post that details all the cool talks that took place at 
+[Cinco de Trino]({% post_url 2022-05-17-cinco-de-trino-recap %}) and
+includes video resources. This was a mini version of the Trino Summit which will
+take place later this year.
+
+includes video resources. This was a mini version of the Trino Summit which will
```
Comment (mosabua) [inline _episodes/36.md:96]: ?? remove..

**Review by mosabua — CHANGES_REQUESTED** [review]
Some minimal clean up left

**Review by mosabua — APPROVED** [review]
This looks good now. Can you please merge @martint ?

**Comment by mosabua** [comment]
Thank you @martint

**Pull request merged**